Fix crash in DiscusMultipleScatteringCorrection

DannyHindson opened the pull request:
**Description of work.**

Some long calculations I was running on PEARL data sets crashed due to an exception being raised in the part of the code that determines the new track direction in between scattering events. This is in the method DiscusMultipleScatteringCorrection::updateTrackDirection
Main problem was that output V3D vector wasn't always of size 1 due to a floating point rounding problem when the input vector was close to vertical (vz~1). This raised an exception in Track::reset.
I've changed `sqrt(1-vz*vz)` to `sqrt(vx*vx + vy*vy)` to avoid subtracting two numbers that are v close together.
Also extend scenarios covered by special case code to cover track that is going vertically down as well as vertically up (ie for vz=-1 as well as vz=1)

**To test:**

I only reproduced the crash in a calculation that took a couple of hours to run. So don't have a nice script that shows the problem before and after. So suggest just doing a code review. The algorithm is covered by various tests on the calculation results in the unit test so the Jenkins checks should ensure these changes haven't changed the results.

Fixes #32988.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Fix couple of issues in updateTrackDirection

Main problem was that output vector wasn't always of size 1 due
to floating point rounding problem when input vector close to vertical.
This raises an exception in Track::reset
Also extend scenarios where special case code for vz=1 to cover vz=-1

DannyHindson committed:
Fix merge conflict in release notes

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -719,27 +719,31 @@ void DiscusMultipleScatteringCorrection::updateTrackDirection(Geometry::Track &t
     UKX = B3 * cos(phi);
     UKY = B3 * sin(phi);
-    UKZ = B2;
+    UKZ = B2 * vz;
```
This seems like a separate bug fix - is it worth mentioning in the release notes?

RichardWaiteSTFC reviewed (commented):
Looks good!
Is it not possible that the same bug could affect the evaluation of the `B3` coefficient as well? I guess we don't tend to evaluate the correction for two-theta -> 0 or 180 deg...
https://github.com/mantidproject/mantid/blob/4e9357de16698ce3583dcb5cc38627af816cd8c2/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp#L711

DannyHindson commented:
> Looks good! Is it not possible that the same bug could affect the evaluation of the `B3` coefficient as well? I guess we don't tend to evaluate the correction for two-theta -> 0 or 180 deg...
> 
> https://github.com/mantidproject/mantid/blob/4e9357de16698ce3583dcb5cc38627af816cd8c2/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp#L711

It's possible this sqrt(1-x*x) does produce a rounding error in a similar way to the one I've fixed but I've not yet seen it cause an exception. And it's the exception that's a bigger deal here than the fact the numbers may be v slightly out. The one I've fixed occasionally caused the size of the new scattering direction vector to be different than 1 which triggers an exception in track.reset. 

Not sure why one sqrt(1-x*x) causes this and the other doesn't - the Rodrigues rotation formula that's implemented in this method must be more sensitive to errors in certain inputs. I'd rather leave this be until it shows up as a problem given the length of time the runs take to test this out.

DannyHindson committed:
Add bug fix relating to tracks pointing exactly down to release notes

RichardWaiteSTFC reviewed (approved):
Thanks for adding the extra release notes, happy to leave the evaluation of `B3` as is, given it is not causing any problems!
